File: catowser/catowser/WebTabs/Site.swift
//
//  Site.swift
//  catowser
//
//  Created by Andrei Ermoshin on 01/02/2019.
//  Copyright © 2019 andreiermoshin. All rights reserved.
//

import Foundation

public struct Site {
    /// Initial url
    public let url: URL

    public var host: String {
        // For http://www.opennet.ru/opennews/art.shtml?num=50072
        // it should be "www.opennet.ru"
        // Add parsing of host https://tools.ietf.org/html/rfc1738#section-3.1
        // in case if iOS sdk returns nil
        return url.host ?? "site"
    }

    /// String associated with site if site was created from search engine.
    /// This convinient property to transfer/save search query to use it for search view.
    /// Different approach could be to store it in tab content type `.site` state as 2nd associated value.
    public let searchSuggestion: String?

    public init(url: URL, searchSuggestion: String? = nil) {
        self.url = url
        self.searchSuggestion = searchSuggestion
    }

    public init?(urlString: String) {
        guard let decodedUrl = URL(string: urlString) else {
            return nil
        }
        url = decodedUrl
        searchSuggestion = nil
    }
}

extension Site: Equatable {}

File: CoreHttpKit/src/commonMain/kotlin/Site.kt
package org.cottonweb.CoreBrowser

import org.cottonweb.CoreHttpKit.Host
import org.cottonweb.CoreHttpKit.URLInfo
import kotlin.native.concurrent.freeze

/**
 *
 * @property urlInfo An initial URL
 * @property faviconData Used by top sites by loading high quality image from the Assets
 * @property searchSuggestion String associated with site if site was created from search engine.
 * This convenient property to transfer/save search query to use it for search view.
 * Different approach could be to store it in tab content type `.site` state as 2nd associated value.
 * */
class Site(
    val urlInfo: URLInfo,
    val settings: Settings,
    internal val faviconData: ByteArray? = null,
    val searchSuggestion: String? = null,
    val userSpecifiedTitle: String? = null
) {

    val host: Host
        get() = urlInfo.host()

    val title: String
        get() {
            if (searchSuggestion != null) {
                return searchSuggestion
            } else if (userSpecifiedTitle != null) {
                return userSpecifiedTitle
            } else {
                return urlInfo.domainName.rawString
            }
        }

    val searchBarContent: String
        get() = searchSuggestion ?: urlInfo.urlWithoutPort

    /**
     * Site settings.
     * */
    data class Settings(
        val isPrivate: Boolean = false,
        val blockPopups: Boolean = true,
        val isJSEnabled: Boolean = true,
        val canLoadPlugins: Boolean = true
    ) {
        fun withChanged(javaScriptEnabled: Boolean): Settings {
            return Settings(isPrivate, blockPopups, javaScriptEnabled, canLoadPlugins)
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            other as Settings
            return isPrivate == other.isPrivate
                    && blockPopups == other.blockPopups
                    && isJSEnabled == other.isJSEnabled
                    && canLoadPlugins == other.canLoadPlugins
        }
    }
}

expect class Image

expect fun Site.withFavicon(image: Image): Site?
expect fun Site.favicon(): Image?
